### DISCUSSION TITLE
Turbo für open3e

### DISCUSSION THREAD

**Discussion by MyHomeMyData** [open3e/open3e#69] [Ideas]
### Massive Laufzeitersparnis durch Weglassen der Dids-Konfig-Prüfung in udsoncan

Aus der Diskussion mit User Wwi im Viessmann-Forum haben wir ja gelernt, dass open3e auf einem Pi Zero kaum vernünftig zu benutzen ist, weil jede Datenpunktabfrage ca. 2 Sekunden dauert. Die allermeiste Zeit wir in der UDSonCAN Lib. mit der Prüfung der Dids-Konfiguration verbraten.

Ich habe eine Version gebastelt, die diese Prüfung schlicht weglässt. Ergebnis: **Bei generischen Abfragen, also ohne devices.json, spart das ca. 80% der Laufzeit auf meinem Pi 4. Bei spezifischen Abfragen mit devices.json ca. 70%.**

Umgesetzt habe ich das, indem ich die fragliche Methode `check_did_config` überladen habe. Damit das funktioniert, musste ich auch die darüber liegenden Klassen überladen. Das steckt alles in der neuen Datei `Open3EudsClient.py`. In 'Open3Eclass.py' ändern sich nur 2 Zeilen. Die neue Lib wird importiert und die Definition des Clients (Zeile 129/130) angepasst. Man kann leicht auf die alte Version umschalten, indem man Zeile 129 verwendet und 130 auskommentiert. Ist nur grob getestet.

Bitte mal ausprobieren, ich habe einen Branch `experimental` angelegt.

Ist das praktikabel?

Nachteile:
* Änderungen an den überladenen Routinen in udsoncan bekommen wir nicht mehr mit
* Schlimmstenfalls funktioniert das Programm nicht mehr, falls sich da Schnittstellen ändern

Die beste Lösung wäre, wir könnten die Kollegen von udsoncan dazu bringen, dass sie einen Schalter einbauen, um die Prüfung der Konfiguration abzuschalten.

**Comment by Wolfgang-03** [comment]
Mir ist die Laufzeitproblematik heute auch aufgefallen. Nachdem ich mein Livesystem heute auf die aktuelle Version von Open3E upgedatet hatte, brauchte ein Scann von 70 DIDs ca. 13 Sekunden, was früher 2 Sekunden und nach der Umstellung auf "udsoncan" in 3 Sekunden gedauert hat. 

Ich habe jetzt eben mal auf "experimental" umgestellt und damit geht das auf immerhin schon mal 6 Sekunden wieder runter. Allerdings sind da auch noch zusätzliche Ausgaben "total time" enthalten.

**Comment by MyHomeMyData** [comment]
@Wolfgang-03 Die Version ist auch nur zum Testen gedacht. Keine Ahnung, ob wir diesen Ansatz weiterverfolgen.

Wenn Dich die Zeitausgaben stören: In `Open3Eclass.py` Zeile 150 auskommentieren.

**Reply by Wolfgang-03** [reply]
Ja, schon klar. Ich teste ja auch nur. Habs mal auskommentiert, auch start_time und end_time, hat aber keine erkennbaren Auswirkungen auf die Laufzeit. Ist bei mir immer noch bei 6 Sekunden für 70 DIDs.

**Reply by philippoo66** [reply]
mit der neuesten udsoncan Version sollte das deepcopy Issue beseitigt sein?! https://github.com/pylessard/python-udsoncan/pull/205 siehe dazu auch https://github.com/pylessard/python-udsoncan/issues/185#issuecomment-1916740996

**Reply by MyHomeMyData** [reply]
Sehr gut! Sobald das per `pip install` verfügbar ist, passen wir die requirements.txt an und das Thema ist erledigt.
